black.py

```diff
@@ -3691,57 +3793,125 @@ def __str__(self) -> str:
         return ", ".join(report) + "."
 
 
-def parse_ast(src: str) -> Union[ast.AST, ast3.AST, ast27.AST]:
-    filename = "<unknown>"
-    if sys.version_info >= (3, 8):
-        # TODO: support Python 4+ ;)
-        for minor_version in range(sys.version_info[1], 4, -1):
-            try:
-                return ast.parse(src, filename, feature_version=(3, minor_version))
-            except SyntaxError:
-                continue
-    else:
-        for feature_version in (7, 6):
-            try:
-                return ast3.parse(src, filename, feature_version=feature_version)
-            except SyntaxError:
-                continue
+def parse_ast(src: str, mode: FileMode) -> AST:
+    """
+    Parse `src` to ast using `typed_ast` or `ast`.
+
+    - uses `mode.target_versions` from the latest to the oldest
+    - if `mode.target_versions` is empty - `py34`, `py35`, `py36`, `py37`
+      and system version
+    - for `py27` uses `typed_ast.ast27`, fails if it is not installed
+    - for `py33` uses `typed_ast.ast3`, fails if it is not installed
+    - for `py38` uses `ast` if system version is 3.8+
+    - for `py38` uses `typed_ast.ast3` if system version is <3.8,
+      fails if it is not installed
+    - for other versions uses `typed_ast.ast3`, fallback to `ast`
+      if `typed_ast is not installed, fallback fails if `typed_ast` is not installed
+
+    Raises:
+        SyntaxError -- If src parse failed for all target versions
+        ModuleNotFoundError -- If `typed_ast` is required but missing or
+            Python 3.8+ required
+    """
+    system_version = TargetVersion.get_sys_version()
+    versions = mode.target_versions
+
+    # all versions that are safe to parse and system if it is Python 3.8+
+    if not versions:
+        versions = {
+            TargetVersion.PY34,
+            TargetVersion.PY35,
+            TargetVersion.PY36,
+            TargetVersion.PY37,
+            TargetVersion.get_sys_version(),
+        }
+
+    # try to parse with current version first
+    versions_sorted = list(sorted(versions, reverse=True))
```

Comment:
sorted() already returns a list. And why do you sort it again on the next line?

Reply:
You are right, conversion to lsit is not needed. And I added more meaningful comments.

```python
    # sort versions from the latest to the oldest
    versions_sorted = sorted(versions, reverse=True)

    # put sys version to the first place to parse source with it first
    versions_sorted.sort(key=lambda v: v != system_version)
```